be/src/exprs/vectorized/math_functions.cpp

```diff
@@ -405,6 +409,95 @@ std::string MathFunctions::decimal_to_base(int64_t src_num, int8_t dest_base) {
     return std::string(buf + max_digits - result_len, result_len);
 }
 
+int128_t MathFunctions::decimal_truncate(const int128_t& lv, const int32_t& l_scale, const int32_t& rv,
+                                         bool keep_scale) {
+    int32_t dec = rv;
+    if (dec < 0) {
+        dec = 0;
+    }
+    int32_t max_precision = decimal_precision_limit<int128_t>;
+    if (dec > max_precision) {
+        dec = max_precision;
+    }
+    int32_t scale_diff = dec - l_scale;
+    int128_t result;
+    if (scale_diff > 0) {
+        if (keep_scale) {
+            result = lv;
+        } else {
+            DecimalV3Cast::to_decimal_truncate<int128_t, int128_t, int128_t, true, true>(
+                    lv, get_scale_factor<int128_t>(scale_diff), &result);
+        }
+    } else if (scale_diff < 0) {
+        DecimalV3Cast::to_decimal_truncate<int128_t, int128_t, int128_t, false, true>(
+                lv, get_scale_factor<int128_t>(-scale_diff), &result);
+        if (keep_scale) {
+            int128_t new_result;
+            DecimalV3Cast::to_decimal_truncate<int128_t, int128_t, int128_t, true, true>(
+                    result, get_scale_factor<int128_t>(-scale_diff), &new_result);
+            result = new_result;
+        }
+    } else {
+        result = lv;
+    }
+    return result;
+}
+
+ColumnPtr MathFunctions::truncate_decimal128(FunctionContext* context, const starrocks::vectorized::Columns& columns) {
+    const auto& type = context->get_return_type();
+
+    ColumnPtr maybe_nullable_c1 = columns[0];
```

Comment:
suggest to use c0, c1 instead of maybe_nullable_{c1, c2}, a ColumnPtr has four cases: const null, const, nullable, not nullable. it's a common sense.

Reply:
Ok. I got it.